src/Components/Components/src/Routing/Router.cs

```diff
@@ -194,51 +195,85 @@ internal virtual void Refresh(bool isNavigationIntercepted)
             if (!OnNavigateAsync.HasDelegate)
             {
                 return true;
             }
 
-            // If we've already invoked a task and stored its CTS, then
-            // cancel that existing CTS.
+            // Cancel the CTS instead of disposing it, since disposing does not
+            // actually cancel and can cause unintended Object Disposed Exceptions.
+            // This effectivelly cancels the previously running task and completes it.
             _onNavigateCts?.Cancel();
-            // Then make sure that the task has been completed cancelled or
-            // completed before continuing with the execution of this current task.
+            // Then make sure that the task has been completely cancelled or completed
+            // before starting the next one. This avoid race conditions where the cancellation
+            // for the previous task was set but not fully completed by the time we get to this
+            // invocation.
             await previousOnNavigate;
 
-            // Create a new cancellation token source for this instance
             _onNavigateCts = new CancellationTokenSource();
             var navigateContext = new NavigationContext(path, _onNavigateCts.Token);
 
-            // Create a cancellation task based on the cancellation token
-            // associated with the current running task.
             var cancellationTcs = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
             navigateContext.CancellationToken.Register(state =>
                 ((TaskCompletionSource)state).SetResult(), cancellationTcs);
 
             var task = OnNavigateAsync.InvokeAsync(navigateContext);
 
-            // If the user provided a Navigating render fragment, then show it.
             if (Navigating != null && task.Status != TaskStatus.RanToCompletion)
             {
                 _renderHandle.Render(Navigating);
             }
 
             var completedTask = await Task.WhenAny(task, cancellationTcs.Task);
-            return task == completedTask;
+
+            // We should not render the component associated with the route we
+            // are navigating to if:
+            // - The OnNavigateAsync task threw an exception
+            // - The OnNavigateAsync task wasn't cancelled by our CTS
+            // - The cancellation task completes first
+            if (task == completedTask)
+            {
+                if (completedTask.IsCompletedSuccessfully)
+                {
+                    return true;
+                }
+
+                try
+                {
+                    await completedTask;
+                }
+                catch (Exception ex)
+                {
+                    if (completedTask.IsCanceled && !_onNavigateCts.IsCancellationRequested)
+                    {
+                        throw new InvalidOperationException("OnNavigateAsync callback cannot be canceled.", ex);
```

Comment:
This [thread](https://github.com/dotnet/aspnetcore/pull/23947#discussion_r455664635) has a little bit more info. We want to throw a specific warning to developers when the ONavigateAsync callback is cancelled, but not by our cancellation token (e.g. the user returns a task that they cancel themselves).

We prohibit this behavior because users might misinterpret canceling the `OnNavigateAsync` task as cancelling the navigation event which isn't possible.

I wanted to explicitly return a IOE to inform the developer that they cannot cancel the callback themselves. Only the framework will handle the responsibility of canceling a task if another one comes in.